crates/matrix-sdk/src/room_preview.rs

```diff
@@ -219,6 +250,105 @@ impl RoomPreview {
 
         let state = client.get_room(room_id).map(|room| room.state());
 
-        Ok(Self::from_room_info(room_info, num_joined_members, state))
+        Ok(Self::from_room_info(client, room_info, num_joined_members, state).await)
     }
 }
+
+pub enum RoomPreviewActions {
+    Invited { join: JoinRoomPreviewAction, leave: LeaveRoomPreviewAction },
+    Knocked { leave: LeaveRoomPreviewAction },
+}
+
+pub struct JoinRoomPreviewAction {
+    client: Client,
+}
+
+impl JoinRoomPreviewAction {
+    pub fn new(client: Client) -> Self {
+        Self { client }
+    }
+
+    pub async fn run(&self, room_preview: &RoomPreview) -> Result<(), RoomPreviewError> {
+        if room_preview.state.is_none() || matches!(room_preview.state, Some(RoomState::Invited)) {
+            match self.client.join_room_by_id(&room_preview.room_id).await {
+                Ok(room) => {
+                    if room_preview.is_direct.unwrap_or(false) {
+                        if let Some(e) = room.set_is_direct(true).await.err() {
+                            warn!("error when setting room as direct: {e}");
+                        }
+                    }
+                    Ok(())
+                }
+                Err(e) => Err(e.into()),
+            }
+        } else {
+            Err(RoomPreviewError::WrongRoomPreviewState(WrongRoomPreviewState::new(
+                "None or Invited",
+                room_preview.state,
+            )))
+        }
+    }
+}
+
+pub struct LeaveRoomPreviewAction {
+    client: Client,
+}
+
+impl LeaveRoomPreviewAction {
+    pub fn new(client: Client) -> Self {
+        Self { client }
+    }
+
+    pub async fn run(&self, room_preview: &RoomPreview) -> Result<(), RoomPreviewError> {
+        let Some(state) = room_preview.state else {
+            return Err(RoomPreviewError::WrongRoomPreviewState(WrongRoomPreviewState::new(
+                "Known room state (not None).",
+                None,
+            )));
+        };
+
+        match state {
+            RoomState::Left => Err(RoomPreviewError::WrongRoomPreviewState(WrongRoomPreviewState::new(
+                "Joined, Invited or Knocked",
+                Some(state),
+            ))),
+            _ => {
+                let request = leave_room::v3::Request::new(room_preview.room_id.to_owned());
+                self.client.send(request, None).await.map_err(Into::into)?;
+                self.client.base_client().room_left(&room_preview.room_id).await.map_err(Into::into)?;
+                Ok(())
+            }
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum RoomPreviewError {
+    /// Attempted to call a method on a room preview that requires the user to
+    /// have a specific membership state, but the membership state is
+    /// different.
+    #[error("wrong room state: {0}")]
+    WrongRoomPreviewState(WrongRoomPreviewState),
+
+    #[error(transparent)]
+    Generic(#[from] Box<dyn std::error::Error + Send + Sync>)
```

Comment:
I really don't like this; having an exceptional generic error type like that should be limited at the strict minimum, i.e. `matrix_sdk::Error` only. Let's avoid it entirely by not making a custom error, but instead use an `matrix_sdk::Error` instead of this error type, there's already `WrongRoomState` which describes the same error.

Reply:
Removed in https://github.com/matrix-org/matrix-rust-sdk/commit/0fcfd84311e2047825763aeae41cc210d3007041.